src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/CSharpConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,301 @@
+                        .Where(statement => semanticModel.GetEnclosingSymbol(statement.SpanStart, cancellationToken) == memberDeclarationSymbol).Count();
+
+                    if (forEachInfo.ForEachStatement.IsParentKind(SyntaxKind.Block) && forEachInfo.ForEachStatement.Parent.Parent == memberDeclarationSyntax)
+                    {
+                        // Check that 
+                        // a. There are either just a single 'yield return' or 'yield return' with 'yield break' just after.
+                        // b. Those foreach and 'yield break' (if exists) are last statements in the method (do not count local function declaration statements).
+                        var statementsOnBlockWithForEach = ((BlockSyntax)forEachInfo.ForEachStatement.Parent).Statements
+                            .Where(statement => statement.Kind() != SyntaxKind.LocalFunctionStatement).ToArray();
+                        var lastNonLocalFunctionStatement = statementsOnBlockWithForEach.Last();
+                        if (yieldStatementsCount == 1 && lastNonLocalFunctionStatement == forEachInfo.ForEachStatement)
+                        {
+                            converter = new YieldReturnConverter(forEachInfo, (YieldStatementSyntax)statementCannotBeConverted, yieldBreakStatement: null);
+                            return true;
+                        }
+
+                        // foreach()
+                        // {
+                        //    yield return ...;
+                        // }
+                        // yield break;
+                        // end of member
+                        if (yieldStatementsCount == 2 &&
+                            lastNonLocalFunctionStatement.Kind() == SyntaxKind.YieldBreakStatement &&
+                            !lastNonLocalFunctionStatement.ContainsDirectives &&
+                            statementsOnBlockWithForEach[statementsOnBlockWithForEach.Length - 2] == forEachInfo.ForEachStatement)
+                        {
+                            // This removes the yield break.
+                            converter = new YieldReturnConverter(forEachInfo, (YieldStatementSyntax)statementCannotBeConverted, yieldBreakStatement: (YieldStatementSyntax)lastNonLocalFunctionStatement);
+                            return true;
+                        }
+                    }
+
+                    break;
+            }
+
+            converter = default;
+            return false;
+        }
+
+        protected override SyntaxNode AddLinqUsing(IConverter<ForEachStatementSyntax, StatementSyntax> converter, SemanticModel semanticModel, SyntaxNode root)
+        {
+            var namespaces = semanticModel.GetUsingNamespacesInScope(converter.ForEachInfo.ForEachStatement);
+            if (!(namespaces.Any(namespaceSymbol => namespaceSymbol.Name == "Linq" && namespaceSymbol.ContainingNamespace.Name == "System")) &&
+                 root is CompilationUnitSyntax compilationUnit)
```

Comment:
when is Root not a compilation unit?

Reply:
I assume that refactorings should be called over compilation units. If not, we still can convert but I'm not sure we can help with adding usings/imports in the case.

I think we can leave it as is until we get an example of usage out of compilation unit. Please let me know if you have a better idea.

---
In reply to: [195845259](https://github.com/dotnet/roslyn/pull/27130#discussion_r195845259) [](ancestors = 195845259)